### ISSUE TITLE
TestPeriodic fails: len(actions) = 0, expected >= 1

### ISSUE THREAD

**Issue by heyitsanthony** [coreos/etcd#6060]
via semaphore,

```
2016-07-26 19:56:41.103215 N | etcdserver: Starting auto-compaction at revision 1
2016-07-26 19:56:41.103422 N | etcdserver: Finished auto-compaction at revision 1
2016-07-26 19:56:41.242566 N | etcdserver: Starting auto-compaction at revision 13
2016-07-26 19:56:41.242635 N | etcdserver: Finished auto-compaction at revision 13
2016-07-26 19:56:41.374956 N | etcdserver: Starting auto-compaction at revision 25
2016-07-26 19:56:41.375034 N | etcdserver: Finished auto-compaction at revision 25
2016-07-26 19:56:41.791289 N | etcdserver: Starting auto-compaction at revision 25
2016-07-26 19:56:41.791354 N | etcdserver: Finished auto-compaction at revision 25
2016-07-26 19:56:41.947325 N | etcdserver: Starting auto-compaction at revision 1
2016-07-26 19:56:41.947393 N | etcdserver: Finished auto-compaction at revision 1
2016-07-26 19:56:42.134529 N | etcdserver: Starting auto-compaction at revision 13
2016-07-26 19:56:42.134619 N | etcdserver: Finished auto-compaction at revision 13
2016-07-26 19:56:42.278163 N | etcdserver: Starting auto-compaction at revision 25
2016-07-26 19:56:42.278234 N | etcdserver: Finished auto-compaction at revision 25
2016-07-26 19:56:42.709166 N | etcdserver: Starting auto-compaction at revision 25
2016-07-26 19:56:42.709414 N | etcdserver: Finished auto-compaction at revision 25
--- FAIL: TestPeriodic (10.13s)
    compactor_test.go:51: len(actions) = 0, expected >= 1
2016-07-26 19:56:53.284981 N | etcdserver: Starting auto-compaction at revision 25
2016-07-26 19:56:53.285066 N | etcdserver: Finished auto-compaction at revision 25
FAIL
```

**Comment by gyuho** [comment]
@heyitsanthony Happened again?


```
Running unit tests...
ok  	github.com/coreos/etcd/auth	2.785s	coverage: 44.2% of statements
ok  	github.com/coreos/etcd/client	1.710s	coverage: 41.3% of statements
ok  	github.com/coreos/etcd/clientv3	7.113s	coverage: 16.8% of statements
ok  	github.com/coreos/etcd/clientv3/naming	1.399s	coverage: 75.0% of statements
2016-08-11 15:16:05.787061 N | etcdserver: Starting auto-compaction at revision 1
2016-08-11 15:16:05.787227 N | etcdserver: Finished auto-compaction at revision 1
2016-08-11 15:16:05.919183 N | etcdserver: Starting auto-compaction at revision 13
2016-08-11 15:16:05.919251 N | etcdserver: Finished auto-compaction at revision 13
2016-08-11 15:16:06.062111 N | etcdserver: Starting auto-compaction at revision 25
2016-08-11 15:16:06.062183 N | etcdserver: Finished auto-compaction at revision 25
2016-08-11 15:16:06.457836 N | etcdserver: Starting auto-compaction at revision 25
2016-08-11 15:16:06.457909 N | etcdserver: Finished auto-compaction at revision 25
--- FAIL: TestPeriodic (10.12s)
	compactor_test.go:54: len(actions) = 0, expected >= 1
2016-08-11 15:16:16.960200 N | etcdserver: Starting auto-compaction at revision 25
2016-08-11 15:16:16.960258 N | etcdserver: Finished auto-compaction at revision 25
2016-08-11 15:16:17.092411 N | etcdserver: Starting auto-compaction at revision 1
2016-08-11 15:16:17.092488 N | etcdserver: Finished auto-compaction at revision 1
2016-08-11 15:16:17.224574 N | etcdserver: Starting auto-compaction at revision 13
2016-08-11 15:16:17.224639 N | etcdserver: Finished auto-compaction at revision 13
2016-08-11 15:16:17.366894 N | etcdserver: Starting auto-compaction at revision 25
2016-08-11 15:16:17.366959 N | etcdserver: Finished auto-compaction at revision 25
2016-08-11 15:16:17.768511 N | etcdserver: Starting auto-compaction at revision 25
2016-08-11 15:16:17.768580 N | etcdserver: Finished auto-compaction at revision 25
FAIL
coverage: 89.7% of statements
```

https://semaphoreci.com/coreos/etcd/branches/master/builds/760

**Comment by gyuho** [comment]
```
--- FAIL: TestPeriodic (10.25s)
	compactor_test.go:57: len(actions) = 0, expected >= 1
2016-08-27 18:56:34.426012 N | etcdserver: Starting auto-compaction at revision 25
2016-08-27 18:56:34.426086 N | etcdserver: Finished auto-compaction at revision 25
FAIL
```

Happened again? https://semaphoreci.com/coreos/etcd/branches/pull-request-6244/builds/11

**Comment by heyitsanthony** [comment]
https://semaphoreci.com/coreos/etcd/branches/master/builds/1472

**Comment by xiang90** [comment]
Really nothing we can do there. The vm just stopped for 10 seconds:

```
2017-01-23 18:20:03.994234 N | compactor: Starting auto-compaction at revision 1
2017-01-23 18:20:03.994288 N | compactor: Finished auto-compaction at revision 1
2017-01-23 18:20:04.141931 N | compactor: Starting auto-compaction at revision 13
2017-01-23 18:20:04.141993 N | compactor: Finished auto-compaction at revision 13
--- FAIL: TestPeriodic (10.42s)
	compactor_test.go:57: len(actions) = 0, expected >= 1
2017-01-23 18:20:14.704074 N | compactor: Starting auto-compaction at revision 25
2017-01-23 18:20:14.704149 N | compactor: Finished auto-compaction at revision 25
```

There is the missing 10 seconds between last compaction and the previous one. Cannot reproduce on my local evn either.

I suggest to just close this issue. Or increase timeout up to 20 seconds if we really care.

**Comment by heyitsanthony** [comment]
The 10s could be from `rg.Wait(1)` timing out (5s), followed by the timeout in compactable.Wait(1) that's throwing the error (another 5s). I don't think it's the VM.

**Comment by xiang90** [comment]
@heyitsanthony So you think rg.Wait and compactable.Wait actually blocks things? I looked at the code, wait only tries to receive from chan, not sending or waiting.

**Comment by heyitsanthony** [comment]
They're using stream recorders, so there's a 5s timeout: https://github.com/coreos/etcd/blob/master/pkg/testutil/recorder.go#L115. The only claim here is the `rg.Wait(1)` never receives the event it expects, and missing that event can cascade down to the `compactable.Wait(1)` missing whatever event it expects and timing out, and that's 10s. `rg.Wait(1)` probably needs an error check at least.

If the VM hangs for 5s, then it'll also cause this error, but the RecorderStream code is all over the place, so I'd expect more failures like this in other tests.

**Comment by xiang90** [comment]
@heyitsanthony It is strange enough since the event actually sent after 10 second

it prints out this after 10 seconds. So I feel the VM got stuck.

```
2017-01-23 18:20:14.704074 N | compactor: Starting auto-compaction at revision 25
2017-01-23 18:20:14.704149 N | compactor: Finished auto-compaction at revision 25
```

**Comment by xiang90** [comment]
Fixed by #7632 again.
